### PR TITLE
Implemented HatEffect

### PR DESCRIPTION
**Pull request by dastgir**
[//]: # (**********************************)
[//]: # (** Fill in the following fields **)
[//]: # (**********************************)

[//]: # (Note: Lines beginning with syntax such as this one, are comments and will not be visible in your report!)

### Pull Request Prelude

[//]: # (Thank you for working on improving Hercules!)

[//]: # (Please complete these steps and check the following boxes by putting an `x` inside the brackets _before_ filing your Pull Request.)

- [x] I have followed [proper Hercules code styling][code].
- [x] I have read and understood the [contribution guidelines][cont] before making this PR.
- [x] I am aware that this PR will be closed if the above-mentioned criteria are not fulfilled.

### Changes Proposed

[//]: # (Describe at length, the changes that this pull request makes.)
Implemented HatEffect Feature.
It is mostly used in items to give new appearance to players.

**Affected Branches:** 

[//]: # (Master? Slave?)

**Issues addressed:**

[//]: # (Issue Tracker Number if any.)

### Known Issues and TODO List

[//]: # (Insert checklist here)
[//]: # (Syntax: - [ ] Checkbox)

[//]: # (**NOTE** Enable the setting "[√] Allow edits from maintainers." when creating your pull request if you have not already enabled it.)

[cont]: https://github.com/HerculesWS/Hercules/blob/master/CONTRIBUTING.md
[code]: https://github.com/HerculesWS/Hercules/wiki/Coding-Style

### CODE REVIEW

**Comment by HerculesWSAPI** [comment]
This change is [Reviewable](https://reviewable.io/reviews/herculesws/hercules/1965)

**Comment by dastgir** [comment]
@Asheraf @4144 Can you review?

**Comment by idamonli** [comment]
@dastgir 

Do you have time to fix this?

https://github.com/HerculesWS/Hercules/pull/1111

File: src/common/mmo.h [hunk 1]
```diff
@@ -64,7 +64,7 @@
 // 20120307 - 2012-03-07aRagexeRE+ - 0x970
 
 #ifndef PACKETVER
-	#define PACKETVER 20141022
+	#define PACKETVER 20170517
```
Comment (4144) [inline src/common/mmo.h]: default version was changed by mistake?

File: src/map/clif.c [hunk 2]
```diff
@@ -20105,6 +20107,56 @@ void clif_skill_scale(struct block_list *bl, int src_id, int x, int y, uint16 sk
...
+void clif_hat_effect(struct block_list *bl, struct block_list *tbl, enum send_target target)
+{
+#if PACKETVER >= 20150422
+	unsigned char *buf;
+	int len, i;
+	struct map_session_data *sd;
+
+	nullpo_retv(bl);
+
+	sd = BL_CAST(BL_PC, bl);
```
Comment (4144) [inline src/map/clif.c]: sd can be null
need add nullpo_retv(sd);

File: src/map/clif.c [hunk 3]
```diff
@@ -20105,6 +20107,56 @@ void clif_skill_scale(struct block_list *bl, int src_id, int x, int y, uint16 sk
...
+		clif->send(buf, len, tbl, target);
+	} else {
+		clif->send(buf, len, bl, target);
+	}
+
+	aFree(buf);
+#endif
+}
+
+void clif_hat_effect_single(struct block_list *bl, uint16 effectId, bool enable){
```
Comment (4144) [inline src/map/clif.c]: bl can be null, need add nullpo_retv(bl)

**Comment by dastgir** [comment]
@4144 Fixed

**Pull request merged**